(Review) Added --last Activation Flag

bpoole16 opened the pull request:
Added flag  "--last" or "-l" in the CLI that gets the most recent activation id for activation's get, logs and result. Closes https://github.com/apache/incubator-openwhisk/issues/1498

dubeejw commented:
@csantanapr, please add me as reviewer here.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -346,6 +385,11 @@ func init() {
     activationListCmd.Flags().Int64Var(&flags.activation.since, "since", 0, wski18n.T("return activations with timestamps later than `SINCE`; measured in milliseconds since Th, 01, Jan 1970"))
 
     activationGetCmd.Flags().BoolVarP(&flags.common.summary, "summary", "s", false, wski18n.T("summarize activation details"))
+    activationGetCmd.Flags().BoolVarP(&flags.activation.last, "last", "l", false, wski18n.T("retrieves last activation"))
```
Need to add the description string to the externalized strings JSON file.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -114,7 +114,11 @@ var activationGetCmd = &cobra.Command{
         var field string
+        args,lerr := lastFlag(args)
 
+        if lerr != nil {
```
`if args, err := lastFlag(args); err != nil {`

dubeejw commented:
Will need doc updates to info users that this new functionality exists. You may need to create an activation doc file as I do not see one in `https://github.com/apache/incubator-openwhisk/blob/master/docs/`.

Also need tests for everything that is added here.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -164,7 +168,6 @@ var activationGetCmd = &cobra.Command{
                 printJSON(activation)
             }
         }
-
```
Leave the break here.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -176,7 +179,11 @@ var activationLogsCmd = &cobra.Command{
     RunE: func(cmd *cobra.Command, args []string) error {
+        args, lerr := lastFlag(args)
 
+        if lerr != nil {
```
Use `if args, err = lastFlag(args); err != nil {` for all of these.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -114,7 +114,11 @@ var activationGetCmd = &cobra.Command{
+        var err error
 
+        if args, err = lastFlag(args); err != nil {
+          return err
```
@mdeuser, do we want a debug message in each of these if statements stating an error occurred?

mdeuser replied:
The `lastFlag` function seems to have some decent debug that seems sufficient.  That said, when an error occurs, the final failure message to the user should be that this command failed with the failure reason.

mdeuser commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -346,6 +386,11 @@ func init() {
+
+    activationLogsCmd.Flags().BoolVarP(&flags.activation.last, "last", "l", false, wski18n.T("retrieves the last activation"))
+
+    activationResultCmd.Flags().BoolVarP(&flags.activation.last, "last", "l", false, wski18n.T("retrieves the last activation"))
```
There's another issue related to `wsk activation poll` in which a single activation id can be used as the poll target.  The `--last` option would also then apply to the `poll` command

rabbah replied:
That's a nice feature; with poll exiting Once result is ready.

mdeuser commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -224,6 +235,35 @@ var activationResultCmd = &cobra.Command{
+func lastFlag(args []string) ([]string, error) {
+  //Checks to see if there is no ID sent with the last flag
+  //If an ID is given with the last flag then the ID will overide the flag
+  if flags.activation.last && len(args) == 0  {
```
What is behavior when activation id and `--last` are both supplied as cli arguments?  This seems conflicting and should be flagged as an usage error.

bpoole16 replied:
Activation ID will be used if both an ID and --last are given.

mdeuser replied:
This might cause some confusion.  Does the user know which one was selected when conflicting id/--last arguments are used on the same command line.  IMHO, this should be flagged as a usage error and not be permitted.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -254,12 +255,17 @@ func lastFlag(args []string) ([]string, error) {
 
     if len(activations) == 0 {
-      whiskErr := whisk.MakeWskError(errors.New("No activations found"), whisk.EXITCODE_ERR_GENERAL, whisk.DISPLAY_MSG, whisk.DISPLAY_USAGE)
+      errStr := fmt.Sprintf("%s%s %s %s", "clinet.Activation.",funcName,"error:", "No activations found")
```
The user should not be informed of the function name directly through an error message. Such information should be displayed in debug messages. Also, the error message is not very clear to someone that doesn't know the internal workings of the CLI.

bpoole16 replied:
Would something like "No activations found. Be sure wsk activation list contains activations" be more clear? This is simply checking to see if they have any activations and if not alerting the user that they don't.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -236,11 +237,11 @@ var activationResultCmd = &cobra.Command{
+//Param: Brings in []strings from args and the name of the function that is using it (for error feedback)
 //Return: Returns an args([]string) with last ID used or the original args with any errors
-func lastFlag(args []string) ([]string, error) {
+func lastFlag(args []string, funcName string) ([]string, error) {
```
I don't think `funcName` is necessary here. Easier to just put a debug message in each of the if statement for the function calls.

bpoole16 replied:
I originally tried to put each error/debug in each separate function but I wasn't able to get `"Unable to get result for activation: {{.err}}"` to display with the actual error message (it would only display the literal {{.err}}) .

dubeejw commented on docs/actions.md:
````diff
@@ -116,6 +116,17 @@ Review the following steps and examples to create your first JavaScript action.
   44794bd6aab74415b4e42a308d880e5b         hello
   6bf1f670ee614a7eb5af3c9fde813043         hello
   ```
+7. If you just need access your last activation you can use the flag `--last` or `-l` to retrieve your last activation with get, logs and result. Run the following command to get your last activation result.
````
"...need access `to` your..." To many `you` and `yours` in here.

mdeuser replied:
have this step immediately follow after step 5 (both talking about activation results).  suggested language for the text `To access the most recent activation record, activation results or activation logs, use the --last or -l flag.`

mdeuser commented:
Affected commands' usage syntax need to be changed.  With the introduction of the `--last` option, the `ACTIVATION_ID` argument is now optional.

Also, need a check to handle when ACTIVATION_ID and and --last are provided at the same time.

dubeejw commented on docs/actions.md:
````diff
@@ -106,7 +106,18 @@ Review the following steps and examples to create your first JavaScript action.
   ```
 
-6. If you forget to record the activation ID, you can get a list of activations ordered from the most recent to the oldest. Run the following command to get a list of your activations:
+  6. To access the most recent activation record, activation results or activation logs, use the `--last` or `-l` flag. Run the following command to get your last activation result.
````
Looks like there is an extra space in front of the 6.

bpoole16 commented:
@mdeuser This checks the conflict between ID and --last 
`else if flags.activation.last && len(args) == 1 {  //Checks conflict between ID and --last
      whiskErr := whisk.MakeWskError(errors.New("Can't use given ID with --last"), whisk.EXITCODE_ERR_GENERAL, whisk.DISPLAY_MSG, whisk.DISPLAY_USAGE)
      return args, whiskErr
  }
  retu`

mdeuser commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -224,6 +248,35 @@ var activationResultCmd = &cobra.Command{
+      whisk.Debug(whisk.DbgInfo, "Appending most recent activation ID into args\n")
+      args = append(args, activations[0].ActivationID)
+    }
+  } else if flags.activation.last && len(args) == 1 {  //Checks conflict between ID and --last
```
A `wsk activation get --last` command may have a `FILTER_FIELD` value.  In this case, the check for a command conflict will fail this command.

In the error message, qualify the `ID` - either use the usage label `ACTIVATION_ID` or just spell it out.. i.e. "activation ID".  Message update: `When specifying an activation ID, do not use the --last flag`

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -224,6 +248,35 @@ var activationResultCmd = &cobra.Command{
+      args = append(args, activations[0].ActivationID)
+    }
+  } else if flags.activation.last && len(args) == 1 {  //Checks conflict between ID and --last
+      whiskErr := whisk.MakeWskError(errors.New("Can't use given ID with --last"), whisk.EXITCODE_ERR_GENERAL, whisk.DISPLAY_MSG, whisk.DISPLAY_USAGE)
```
Currently we can pass an activation ID and a field filter to the activation get command. However, your change break the field filter functionality.

Field filtering with activation get:
```
$ wsk activation get fa613f7e33dd45959b8a93943ea5c3e7 publish
ok: got activation fa613f7e33dd45959b8a93943ea5c3e7, displaying field publish
false
```

Attempting to field filter with your changes:
```
 $ wsk activation get --last  publish
error: Unable to get result for activation: Can't use given ID with --last
Run 'wsk --help' for usage.
```

bpoole16 commented:
Changed conflict code with ID and --last to work with filter fields, updated docs/comments, updated Activation_ID syntax, and added polling for ID with integration of --last flag.

bpoole16 commented:
PG 4 #445

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -98,14 +98,22 @@ var activationListCmd = &cobra.Command{
+        var err error
 
+        if args, err = lastFlag(args); err != nil {  // Checks if any errors occured in lastFlag(args)
+          whisk.Debug(whisk.DbgError, "client.Activation.Get failed: %s\n", err)
```
Method in debug message should match name of the function called, and add the args to the message.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -161,13 +169,21 @@ var activationGetCmd = &cobra.Command{
+        var err error
 
+        if args, err = lastFlag(args); err != nil {  // Checks if any errors occured in lastFlag(args)
+          whisk.Debug(whisk.DbgError, "client.Activation.Logs failed: %s\n", err)
```
Same as comment above.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -189,13 +205,21 @@ var activationLogsCmd = &cobra.Command{
+        var err error
 
+        if args, err = lastFlag(args); err != nil {  // Checks if any errors occured in lastFlag(args)
+          whisk.Debug(whisk.DbgError, "client.Activation.Result failed: %s\n", err)
```
Same as comment above.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -216,6 +240,47 @@ var activationResultCmd = &cobra.Command{
+    }
+    if len(activations) == 0 {    // Checks to to see if there are activations available
+      whisk.Debug(whisk.DbgError, "No activations found in activation list\n")
+      errStr := wski18n.T("Activation list contains no activations")
```
Maybe something like: `Activation list does not contain any activations.`.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -216,6 +240,47 @@ var activationResultCmd = &cobra.Command{
+      return args, whiskErr
+    } else {
+      if len(args) == 0 {
+        whisk.Debug(whisk.DbgInfo, "Appending most recent activation ID into args\n")
```
Should probably print the arguments you are appending in the debug messages in this function.

bpoole16 replied:
I could do a for loop or something checking the length of the arguments to see if they are an ID (length of 32) or filter field (any length).

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -216,6 +240,47 @@ var activationResultCmd = &cobra.Command{
+func lastFlag(args []string) ([]string, error) {
+  // Checks to see if there is an ID sent with the --last
+  // If an ID is given with --last then an error will be thrown
+  if  flags.activation.last {
```
Remove a space after `if`.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -216,6 +240,47 @@ var activationResultCmd = &cobra.Command{
+    }
+    activations,_, err := client.Activations.List(options)
+    if err != nil {   // Checks Activations.List for errors when retrieving latest activaiton
+      whisk.Debug(whisk.DbgError, "client.Activations.List() error for flag --last: %s\n", err)
```
Show arguments passed to the method in debug statements.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -216,6 +240,47 @@ var activationResultCmd = &cobra.Command{
+      if len(args) == 0 {
+        whisk.Debug(whisk.DbgInfo, "Appending most recent activation ID into args\n")
+        args = append(args, activations[0].ActivationID)
+      } else if _,_,err := client.Activations.Get(args[0]); len(args) == 1 && err != nil{   // Checks to see if filter field is given and not ID
```
Looks like maybe you should check the arg length before making calls to the server.

bpoole16 replied:
I could do a for loop or something checking the length of the arguments to see if they are an ID (length of 32) or filter field (any length).

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -216,6 +240,47 @@ var activationResultCmd = &cobra.Command{
+          args = append(args, activations[0].ActivationID)
+          whisk.Debug(whisk.DbgInfo, "Allocating appended ID to correct position in args\n")
+          args[0], args[1] = args[1], args[0]   // IDs should be located at args[0], if a filter field is given it has to be moved to args[0]
+      } else if _,_,err := client.Activations.Get(args[0]);  len(args) >= 1 && err == nil {    // Checks conflict between ID, field filter and --last
```
You should probably check to see if the last flag and an activation ID exists in the args before you do anything of this processing.

bpoole16 replied:
I could do a for loop or something checking the length of the arguments to see if they are an ID (length of 32) or filter field (any length). The point of the system calls were to check if they were returning errors due to filter field being sent instead of ID.

rabbah commented:
for all the style related comments, https://github.com/apache/incubator-openwhisk/tree/master/tools/eclipse has an eclipse code formatting style. maybe we need automatic style checks now.

bpoole16 commented on docs/actions.md:
````diff
@@ -887,9 +898,7 @@ You can use the OpenWhisk CLI to watch the output of actions as they are invoked
   wsk activation poll
   ```
-
-  This command starts a polling loop that continuously checks for logs from activations.
````
Looks like it got deleted when I was removed polling + --last doc stuff, will add back.

bpoole16 commented:
PG #821

dubeejw commented on tests/src/test/scala/common/Wsk.scala:
```diff
@@ -557,14 +557,19 @@ class WskActivation()
-        val params = { fieldFilter map { f => Seq(f) } getOrElse Seq() }
-        cli(wp.overrides ++ Seq(noun, "get", "--auth", wp.authKey, activationId) ++ params, expectedExitCode)
+        val params =
+          { activationId map {a => Seq(a)} getOrElse Seq() } ++
```
Space needed after `activationId map {`, and after `Seq(a)`.

dubeejw commented on tests/src/test/scala/common/Wsk.scala:
```diff
@@ -573,12 +578,17 @@ class WskActivation()
             implicit wp: WskProps): RunResult = {
-        cli(wp.overrides ++ Seq(noun, "logs", activationId, "--auth", wp.authKey), expectedExitCode)
+        val params =
+          { activationId map {a => Seq(a)} getOrElse Seq() } ++
```
Space needed after `activationId map {`, and after `Seq(a)`.

dubeejw commented on tests/src/test/scala/common/Wsk.scala:
```diff
@@ -587,12 +597,17 @@ class WskActivation()
             implicit wp: WskProps): RunResult = {
-        cli(wp.overrides ++ Seq(noun, "result", activationId, "--auth", wp.authKey), expectedExitCode)
+        val params =
+          { activationId map {a => Seq(a)} getOrElse Seq() } ++
```
Space needed after `activationId map {`, and after `Seq(a)`.

dubeejw commented on tests/src/test/scala/common/Wsk.scala:
```diff
@@ -587,12 +597,17 @@ class WskActivation()
-        cli(wp.overrides ++ Seq(noun, "result", activationId, "--auth", wp.authKey), expectedExitCode)
+        val params =
+          { activationId map {a => Seq(a)} getOrElse Seq() } ++
+          { last map {l => Seq("--last")} getOrElse Seq() }
```
Space needed after `last map {`, and after `Seq("--last")`.

dubeejw commented on tests/src/test/scala/common/Wsk.scala:
```diff
@@ -573,12 +578,17 @@ class WskActivation()
-        cli(wp.overrides ++ Seq(noun, "logs", activationId, "--auth", wp.authKey), expectedExitCode)
+        val params =
+          { activationId map {a => Seq(a)} getOrElse Seq() } ++
+          { last map {l => Seq("--last")} getOrElse Seq() }
```
Space needed after `last map {`, and after `Seq("--last")`.

dubeejw commented on tests/src/test/scala/common/Wsk.scala:
```diff
@@ -557,14 +557,19 @@ class WskActivation()
+        val params =
+          { activationId map {a => Seq(a)} getOrElse Seq() } ++
+          { fieldFilter map { f => Seq(f) } getOrElse Seq() } ++
+          { last map {l => Seq("--last")} getOrElse Seq() }
```
Space needed after `last map {`, and after `Seq("--last")`.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -101,14 +101,22 @@ var activationListCmd = &cobra.Command{
+        var err error
 
+        if args, err = lastFlag(args); err != nil {  // Checks if any errors occured in lastFlag(args)
+          whisk.Debug(whisk.DbgError, "client.Activation.Get(%#v) failed: %s\n", args, err)
```
Debug message should contain `lastFlag` instead of `client.Activation.Get` since `lastFlag` is the method you are calling, and allows us to trace the code execution path through debug messages. Same applies to the other instances were `lastFlag` is called below.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -219,6 +243,45 @@ var activationResultCmd = &cobra.Command{
+                  args = append(args, activations[0].ActivationID)
+                  whisk.Debug(whisk.DbgInfo, "Allocating appended ID to correct position in args\n")
+                  args[0], args[1] = args[1], args[0]    // IDs should be located at args[0], if a filter field is given it has to be moved to args[0]
+            } else if len(args[0]) == 32 {    // Checks conflict between ID and --last
```
Same as above if the field filter is 32 characters long, this doesn't work.

dubeejw replied:
Maybe better to prepend the retrieved activation ID to the args always. I think we will only have an issue if a user specified an activation ID with the `--last` flag. Under that scenario they user would get an invalid field filter message. However, the usage says not to include an activation ID with the `--last` flag. @mdeuser, what do you think?

bpoole16 replied:
Not sure how to distinguish them then, was originally using `_,_,err := client.Activations.Get(args[0])` to see if it was an ID and if it wasn't an ID it would return with an error.

bpoole16 replied:
Found a way to use the fieldExists method in util to check if the args[0] is holding a filter field or not. Seems like a sound way to get around this issue.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -219,6 +243,45 @@ var activationResultCmd = &cobra.Command{
+            if len(args) == 0 {
+                whisk.Debug(whisk.DbgInfo, "Appending most recent activation ID(%s) into args\n", activations[0].ActivationID)
+                args = append(args, activations[0].ActivationID)
+            } else if len(args) == 1 && len(args[0]) != 32 {    // Checks to see if filter field is given and not ID
```
What if there is a field filter that is 32 characters long?

bpoole16 commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -262,24 +262,26 @@ func lastFlag(args []string) ([]string, error) {
+            args = append(args, activations[0].ActivationID)
+        } else {
+			filterExists := fieldExists(&whisk.Activation{}, args[0])
+			if filterExists == true {    // Checks to see if filter field is given and not ID
```
Not sure why it is doing this, Doesn't look like this in my IDE.

mdeuser replied:
@bpoole16 - be sure your ide translates tabs into spaces.

dubeejw commented on tools/cli/go-whisk-cli/commands/activation.go:
```diff
@@ -219,6 +243,47 @@ var activationResultCmd = &cobra.Command{
+                args[0], args[1] = args[1], args[0]    // IDs should be located at args[0], if a filter field is given it has to be moved to args[0]
+            } else if filterExists == false {    // Checks conflict between ID and --last
+                whisk.Debug(whisk.DbgError,"Can not use activation ID with --last flag\n")
+                errStr := wski18n.T("When specifying an activation ID, do not use the --last flag.")
```
This isn't necessarily true when an actual field filter is supplied for a filed that doesn't exist. Ex:
```
wsk activation get --last bogusField
```

dubeejw replied:
Can probably just drop the field filtering checks as the actual command should handle this check.

dubeejw replied:
Also `logs` and `result` doesn't support a field filter.

bpoole16 replied:
I'm using the filter field to check to see if it is an ID or a filter field, without it when you use `wsk activation get ACTIVATION_ID --last FIELD_FILTER` and it will use the activation_id given over the flag `--last`. If we are just summing that up to user error then I don't need the check otherwise it will need to check and see if an ID or field filter is given.

dubeejw replied:
I am incline to let to say that is a user error and not provide error handling for it as there is no way to tell the difference between an activation ID and a field filter.

bpoole16 replied:
Works for me.

dubeejw commented:
@bpoole16, you may want to open an issue that states that we could move the field filter to a flag. With that functionality we will be able to tell that the field filter is not the activation ID.  See examples below:

```
wsk activation get cc5ea085c7824846959269f4a4c14e6f --fieldFilter response
```


```
wsk activation get--last --fieldFilter response
```

dubeejw commented on tools/cli/build.gradle:
```diff
@@ -38,9 +38,12 @@ task copyCLIShortcut(type: Copy, dependsOn: [distBinary, dumpOSInfo]) {
-    from from_path_wsk
-    into to_path_dir
+    if (file(from_path_wsk).exists()) {
+      exec {
```
This exec cal doesn't seem to build the CLI on a mac.

dubeejw replied:
@bpoole16, Remove the changes to the build.gradle and I say we can merge this one after Travis passes. Will resolve your build issues in a different PR.

bpoole16 replied:
Okay, will do.

dubeejw commented:
@bjustin-ibm, please merge.

rabbah commented on tests/src/test/scala/system/basic/WskSequenceTests.scala:
```diff
@@ -216,7 +216,7 @@ class WskSequenceTests
                     // check that inner sequence had only (limit - 1) activations
                     val innerSeq = activation.logs.get(1) // the id of the inner sequence activation
-                    val getInnerSeq = wsk.activation.get(innerSeq)
+                    val getInnerSeq = wsk.activation.get(Some(innerSeq))
```
hmm seems wrapping all ids with Some is too common - should there be an overloaded `get` instead of changing all the test cases?

rabbah replied:
counted only 10... ok maybe not as bad as I thought.
